### ISSUE TITLE
Tagger not working as expected when using get_pipe("tagger")

### ISSUE THREAD

**Issue by DavidHuebner** [explosion/spaCy#7105]
## How to reproduce the behaviour
I want to apply the PoS tagger to a list of predefined Tokens and try:

``` python
import spacy
from spacy.tokens import Doc
nlp = spacy.load("en_core_web_sm")
doc = Doc(nlp.vocab, words=["Berlin","is","great"])
doc = nlp.get_pipe("tagger")(doc)
for token in doc:
    print(f"Token: {token.text:10} PoS: {token.pos_}")
```
This prints:
```
Token: Berlin     PoS: 
Token: is         PoS: 
Token: great      PoS: 
```
All the PoS tags are None, however, `doc.has_annotation("TAG")` is `True`. Also, I get meaningful results if I select the `ner` pipeline instead of the `tagger`. I also obtain meaningful results if I directly apply the model to the text with 
```python
doc = nlp("Berlin is great")
for token in doc:
    print(f"Token: {token.text:10} PoS: {token.pos_}")
```
prints
```
Token: Berlin     PoS: PROPN
Token: is         PoS: AUX
Token: great      PoS: ADJ
```
I don't understand why my initial approach fails.

## Your Environment

- **spaCy version:** 3.0.3
- **Platform:** Linux-5.8.0-43-generic-x86_64-with-glibc2.29
- **Python version:** 3.8.5
- **Pipelines:** en_core_web_sm (3.0.0)

**Comment by svlandeg** [comment]
You've stumbled upon some subtle difference in v3 that is masking the issue ;-)

Basically - your code is largely fine and that's indeed how you can call the `tagger`, BUT the tagger currently only sets `token.tag_`, and not `token.pos_` (cf https://github.com/explosion/spaCy/pull/5848). You'll see that it works if you print `token.tag_` in your code.

Also as a side note - you don't have to assign `doc = ` - just using `nlp.get_pipe("tagger")(doc)` works fine because it modifies the doc inplace.

**Comment by DavidHuebner** [comment]
Thanks for your very quick response! I tested your suggestion:

``` python
import spacy
from spacy.tokens import Doc
nlp = spacy.load("en_core_web_sm")
doc = Doc(nlp.vocab, words=["Berlin","is","great"])
nlp.get_pipe("tagger")(doc)
for token in doc:
    print(f"Token: {token.text:10} PoS: {token.tag_}")
```
prints
```
Token: Berlin     PoS: IN
Token: is         PoS: IN
Token: great      PoS: IN
```
whereas the following code block yields different results.
``` python
import spacy
nlp = spacy.load("en_core_web_sm")
doc = nlp("Berlin is great")
for token in doc:
    print(f"Token: {token.text:10} PoS: {token.tag_}")
```

```
Token: Berlin     PoS: NNP
Token: is         PoS: VBZ
Token: great      PoS: JJ
```

Again, I would expect the outputs to be the same. What am I missing here?

**Comment by svlandeg** [comment]
In spaCy v3, the pipeline components are often not independent anymore like they used to be in v2, so it's expected that running just a single component (while technically possible) will not produce good results perse.

In the case of `en_core_web_sm`, the `tagger` connects to an earlier component in the pipeline called the `tok2vec`. It does this with a listener pattern - more information can be found here: https://spacy.io/usage/embeddings-transformers#embedding-layers

In short, if you'd add 
```
nlp.get_pipe("tok2vec")(doc)
```
to your code, right before you call the tagger, it will produce the same results as when running the full pipeline.

I think it's probably easier to avoid calling the components directly though, and use 
```
nlp.select_pipes(enable=["tok2vec", "tagger"])
```
instead.

**Comment by DavidHuebner** [comment]
Again, thank you very much for that quick response. It is now working as expected.
Just one final question: If I only want to use the `ner` pipeline, is it sufficient to call `nlp.get_pipe("ner")(doc)` or do I also need some preprocessing steps?

**Comment by svlandeg** [comment]
The NER is actually independent in `en_core_web_sm`, so you should be able to run it by itself.

You can check this for yourself by locating the folder where `en_core_web_sm` was installed and navigating to `config.cfg`. For the tagger, you'll see this:
```
[components.tagger.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"
width = ${components.tok2vec.model.encode:width}
upstream = "*"
```
The fact that it uses a `Tok2VecListener` as its `tok2vec` sublayer, means it relies on the `tok2vec` component earlier on in the pipeline.

By contrast, the NER has this bit:
```
[components.ner.model.tok2vec]
@architectures = "spacy.Tok2Vec.v1"
```
Which means that the NER's `tok2vec` component is actually specific to the NER, and included in its model, so the NER can be run in isolation.

Ofcourse in general there may be other ways in which components depend on eachother, so you'd need to know how the components work together before isolating some of them. Also have a look at https://spacy.io/usage/processing-pipelines#analysis which shows you the attributes being set by different components!

**Comment by DavidHuebner** [comment]
Thank you so much! That were the quickest responses in Github Issues that I have ever seen.
